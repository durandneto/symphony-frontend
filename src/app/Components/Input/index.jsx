import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import
  StyledInput,
  {
    StyledIconLoading,
    StyledIconError,
    StyledErrorMessage
  }
from './styles'

const Input = props => {
  const [value, handleChange] = useState(props.value)
  const [inputValid, setInputInvalid] = useState(true)

  useEffect(() =>{
    handleChange(props.value)
    if ( props.test && props.value ) {
      setInputInvalid(!props.value.match(props.test))
    }
  }, [props.value, props.test])

  return (
    <React.Fragment>
      <StyledInput
        id={props.id}
        type={props.type}
        highlight={props.highlight}
        textCenter={props.textCenter}
        placeholder={props.placeholder}
        success={props.success}
        error={props.error || !inputValid}
        disabled={props.disabled || props.loading}
        loading={props.loading || false}
        autoFocus={props.autoFocus}
        value={value}
        onChange={e => {
          handleChange(e.currentTarget.value)
          if (!e.currentTarget.value) {
            setInputInvalid(true)
          } else if (props.test) {
            setInputInvalid(!e.currentTarget.value.match(props.test))
          }
        }}
        onBlur={e => {
          inputValid && props.onEvent({
            data: {
              value,
              data: props.data
            },
            event: "onBlur",
            origin: "Input"
          })
        }}

        onFocus={e => {
          inputValid && props.onEvent({
            data: {
              value,
              data: props.data
            },
            event: "onFocus",
            origin: "Input"
          })
        }}

        onKeyUp={e => {
          if (e.key === "Enter") {
            inputValid && props.onEvent({
              data: {
                value,
                data: props.data
              },
              event: "onKeyUpAction",
              origin: "Input"
            })
          } else {
            // handleChange(e.currentTarget.value)
            inputValid && props.onEvent({
                data: {
                  value,
                  data: props.data
                },
                event: "onKeyUp",
                origin: "Input"
              })
            }
        }}
    />
    {
      props.loading && <StyledIconLoading />
    }
    {
      (props.error || !inputValid)  &&
        <React.Fragment>
          <StyledIconError />
          <StyledErrorMessage highlight={props.highlight}>
            {
              !inputValid ? props.testMessage :  props.error
            }
          </StyledErrorMessage>
        </React.Fragment>
    }
  </React.Fragment>
  )
}

Input.defaultProps = {
  type: "text",
  testMessage: "",
  disabled: false,
  highlight: false,
  textCenter: false,
  loading: false,
  placeholder: '',
  onEvent: () => {}
}

Input.propTypes = {
  disabled: PropTypes.bool,
  test: PropTypes.any,
  highlight: PropTypes.bool,
  textCenter: PropTypes.bool,
  loading: PropTypes.bool,
  placeholder: PropTypes.string,
  type: PropTypes.string,
  testMessage: PropTypes.string,
  value: PropTypes.string.isRequired,
  onEvent: PropTypes.func,
  data: PropTypes.object,
}

export default Input
