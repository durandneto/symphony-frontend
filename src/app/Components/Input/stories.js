import React from "react";
import { storiesOf } from "@storybook/react";
import Container, { Row, Column } from '../Grid'
import HeaderFont from '../Typography/Header'
import BodyFont from '../Typography/Body'
import Input from ".";


storiesOf("Components|Input", module)
  .add("default", () => (
    <Container>
      <Row>
        <Column full>
          <Input
            value=""
            data={{User: { id: 1}}}
            placeholder="Input"
            onEvent={e => {
              console.log(e)
            }}
          />
        </Column>
      </Row>
    </Container>
  ))
  .add("Variations", () => (
    <Container>
      <Row >
        <Column full>
          <HeaderFont h1 label={"Controlled Inputs"} />
        </Column>
      </Row>
      <Row >
        <Column full>
          <HeaderFont h3 color={"Blue-Grey-500"} label={"No Validation"} />
        </Column>
      </Row>
      <Row>
        <Column full>
          <Input value="" placeholder="Simple Input" onEvent={e => {
            console.log(e)
          }} />
        </Column>
      </Row>
      <Row >
        <Column full>
          <HeaderFont h3 label={"Prioritizing component validation"} />
        </Column>
      </Row>
      <Row >
        <Column full>
          <BodyFont color={"Blue-Grey-300"} label={`Regex /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/`} />
        </Column>
        <Column full>
          <BodyFont color={"Blue-Grey-300"} label={`Regex /^\d*$/`} />
        </Column>
      </Row>
      <Row >
        <Column full>
          <Input
            test={/^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/}
            testMessage={"Invalid email"}
            value=""
            placeholder="Type a email"
            onEvent={e => {
            console.log(e)
          }} />
        </Column>
        <Column full>
        <Input
            test={/^\d*$/}
            testMessage={"Not a number"}
            value=""
            placeholder="Type a number"
            onEvent={e => {
            console.log(e)
          }} />
        </Column>
      </Row>
      <Row >
        <Column full>
          <HeaderFont h3  color={"Red-500"}  label={"Prioritizing Server Validation"} />
        </Column>
      </Row>
      <Row >
        <Column full>
          <BodyFont color={"Blue-Grey-300"} label={`Regex /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/`} />
        </Column>
        <Column full>
          <BodyFont color={"Blue-Grey-300"} label={`Regex /^\d*$/`} />
        </Column>
      </Row>
      <Row >
        <Column full>
          <Input
            test={/^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/}
            testMessage={"Invalid email"}
            error="Server Error 503"
            value="xxx@xxx.com"
            placeholder="Type a email"
            onEvent={e => {
            console.log(e)
          }} />
        </Column>
        <Column full>
        <Input
            test={/^\d*$/}
            testMessage={"Not a number"}
            value="zxczx"
            placeholder="Type a number"
            error="Server Error 500"
            onEvent={e => {
            console.log(e)
          }} />
        </Column>
      </Row>
      <Row >
        <Column full>
          <HeaderFont h3  color={"Blue-500"}  label={"Loading"} />
        </Column>
      </Row>
      <Row>
        <Column full>
          <Input loading value="Simple Input loading" placeholder="Simple Input loading" onEvent={e => {
            console.log(e)
          }} />
        </Column>
        <Column full>
          <Input loading value="Simple Input loading" placeholder="Simple Input loading" onEvent={e => {
            console.log(e)
          }} />
        </Column>
        <Column full>
          <Input loading value="Simple Input loading" placeholder="Simple Input loading" onEvent={e => {
            console.log(e)
          }} />
        </Column>
      </Row>
      <Row>
        <Column full>
          <Input loading value="Simple Input loading" placeholder="Simple Input loading" onEvent={e => {
            console.log(e)
          }} />
        </Column>
      </Row>
      <Row >
        <Column full>
          <HeaderFont h3  color={"Green-500"}  label={"success"} />
        </Column>
      </Row>
      <Row>
        <Column full>
          <Input success value="Simple Input success" placeholder="Simple Input success" onEvent={e => {
            console.log(e)
          }} />
        </Column>
        <Column full>
          <Input success value="Simple Input success" placeholder="Simple Input success" onEvent={e => {
            console.log(e)
          }} />
        </Column>
        <Column full>
          <Input success value="Simple Input success" placeholder="Simple Input success" onEvent={e => {
            console.log(e)
          }} />
        </Column>
      </Row>
      <Row>
        <Column full>
          <Input success value="Simple Input disabled" placeholder="Simple Input disabled" onEvent={e => {
            console.log(e)
          }} />
        </Column>
      </Row>
      <Row >
        <Column full>
          <HeaderFont h3  color={"Blue-Grey-500"}  label={"Disabled"} />
        </Column>
      </Row>
      <Row>
        <Column full>
          <Input disabled value="Simple Input disabled" placeholder="Simple Input disabled" onEvent={e => {
            console.log(e)
          }} />
        </Column>
        <Column full>
          <Input disabled value="Simple Input disabled" placeholder="Simple Input disabled" onEvent={e => {
            console.log(e)
          }} />
        </Column>
        <Column full>
          <Input disabled value="Simple Input disabled" placeholder="Simple Input disabled" onEvent={e => {
            console.log(e)
          }} />
        </Column>
      </Row>
      <Row>
        <Column full>
          <Input disabled value="Simple Input disabled" placeholder="Simple Input disabled" onEvent={e => {
            console.log(e)
          }} />
        </Column>
      </Row>
    </Container>
  ))