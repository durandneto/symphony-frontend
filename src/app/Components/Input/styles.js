import styled, { keyframes } from 'styled-components'
import Colors from '../Colors'

const inputLoadingAnimation = keyframes`
  to {
    transform: rotate(360deg);
  }
`

export const StyledErrorMessage = styled.span`
  color: ${Colors['Red-600']};
  top: 42px;
  ${({ highlight }) => highlight && `
    top: 72px;
  `}
  position: absolute;
  width: 100%;
  overflow: hidden;
  height: 15px;
  font-size: 12px;
`

export const StyledIconError = styled.span`
    color: ${Colors['Red-700']};
    top: 45px;
    position: absolute;
    width: 100%;
    :before {
        content: '!';
        color: #fff ;
        padding-left: 6px;
        box-sizing: border-box;
        position: absolute;
        top: -21px;;
        right: 21px;
        width: 20px;
        height: 20px;
        margin-top: -10px;
        margin-left: -10px;
        border-radius: 50%;
        border: 2px solid #fff;
    }
`

export const StyledIconLoading = styled.span`
  cursor: wait;
  :before {
    content: '';
    box-sizing: border-box;
    position: absolute;
    top: 50%;
    right: 12px;
    width: 20px;
    height: 20px;
    margin-top: -10px;
    margin-left: -10px;
    border-radius: 50%;
    border: 1px solid ${Colors['Blue-Grey-200']};
    border-top-color: ${Colors['Blue-500']};
    animation: ${inputLoadingAnimation} .6s linear infinite;
  }
`

export const Input = styled.input`
  position: relative;
  border: none;
  border-bottom: 1px solid ${Colors["Blue-Grey-100"]};
  color: ${Colors["Grey-900"]};
  font-size: 14px;
  padding: 10px 50px 10px 20px;
  width: 100%;
  box-sizing: border-box;

  ::placeholder {
    font-size: 14px;
    ${({ highlight }) => highlight && `
      font-size: 2.4rem;
    `}
    font-weight: normal;
    font-stretch: normal;
    font-style: normal;
    line-height: 1;
    letter-spacing: normal;
    color: ${Colors["Blue-Grey-100"]};
  }

  ${({ disabled }) => disabled && `
    color: ${Colors["Blue-Gray-200"]};
    border-color: ${Colors["Blue-Gray-200"]};
    cursor: not-allowed;
    ::placeholder {
      color: ${Colors["Blue-Gray-100"]};
    }`}

  ${({ textCenter }) => textCenter && `
    text-align: center;
  `}

  ${({ highlight }) => highlight && `
    font-size: 2.4rem;
  `}

  ${({ loading }) => loading && `
    color: ${Colors["Light-Blue-200"]};
    border-color: ${Colors["Light-Blue-200"]};
    background-color: transparent;
    cursor: wait;
    ::placeholder {
      color: ${Colors["Light-Blue-100"]};
    }`}

  ${({ error }) => error && `
    color: ${Colors["Red-500"]};
    border-color: ${Colors["Red-500"]};
    ::placeholder {
      color: ${Colors["Red-100"]};
    }`}

  ${({ success }) => success && `
    color: ${Colors["Green-500"]};
    border-color: ${Colors["Green-500"]};
    ::placeholder {
      color: ${Colors["Green-100"]};
    }`}

`;

export default Input