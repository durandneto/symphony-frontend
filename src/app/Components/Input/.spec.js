import React from 'react'
import renderer from 'react-test-renderer'
import Input from '.'

describe('Input', () => {
  it ('Rendering Input without crashing', () => {
    const result = renderer.create(
      <Input
        value="Test input"
        data={{User: { id: 1}}}
        placeholder="Input"
        onEvent={e => {
          console.log(e)
        }}
      />
    ).toJSON()
    expect(result).toMatchSnapshot()
  })
})
