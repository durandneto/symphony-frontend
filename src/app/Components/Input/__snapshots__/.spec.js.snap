// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Rendering Input without crashing 1`] = `
<input
  className="sc-AxhCb hkRQIS"
  disabled={false}
  loading={false}
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyUp={[Function]}
  placeholder="Input"
  type="text"
  value="Test input"
/>
`;
